external/implementations/samlconf-openam-impl/src/main/java/org/codice/compliance/saml/plugin/openam/OpenAMIdpSSOResponderProvider.java

```diff
@@ -0,0 +1,166 @@
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import org.codice.compliance.saml.plugin.IdpSSOResponder;
+import org.codice.compliance.utils.ConstantsKt;
+import org.codice.compliance.utils.GlobalSession;
+
+public class OpenAMIdpSSOResponderProvider implements IdpSSOResponder {
+
+  private static String TOKEN_ID = "tokenId";
+  private OpenAMDataParser openAMData = new OpenAMDataParser();
+
+  @Override
+  public Response getResponseForRedirectRequest(Response originalResponse) {
+    return openAMLogin(originalResponse);
+  }
+
+  @Override
+  public Response getResponseForPostRequest(Response originalResponse) {
+    return openAMLogin(originalResponse);
+  }
+
+  private Response openAMLogin(Response res) {
+
+    //  The Authentication Realm to connect to
+    String authURL =
+        openAMData.getBaseURL()
+            + "/json/realms/root"
+            + openAMData.getRealmsUrl()
+            + openAMData.getOpenAMRealm()
+            + "/authenticate";
+
+    Map<String, String> queryParams = getQueryParams(res.getHeader(ConstantsKt.LOCATION));
+
+    // -- Get the initial json form to fill in --
+    JsonParser parser = new JsonParser();
+    Response payloadRes = RestAssured.given().queryParams(queryParams).when().post(authURL);
+
+    JsonObject payload = parser.parse(payloadRes.jsonPath().prettify()).getAsJsonObject();
+
+    JsonArray cbArr = payload.get("callbacks").getAsJsonArray();
+
+    cbArr.forEach(this::completeForm);
+
+    // -- send the completed form to get authenticated --
+    Response authRes =
+        RestAssured.given()
+            .when()
+            .body(payload)
+            .contentType("application/json")
+            .log()
+            .ifValidationFails()
+            .queryParams(queryParams)
+            .when()
+            .post(authURL);
+
+    JsonObject successJson = parser.parse(authRes.jsonPath().prettify()).getAsJsonObject();
+
+    if (!successJson.has(TOKEN_ID)) {
+      throw new SecurityException(
+          "Failed login due to invalid credentials. "
+              + "Check the SamlCTK OpenAM documentation for more information. "
+              + openAMData.getFormMapping().toString());
+    }
+
+    Map<String, String> cookies = new HashMap<>();
+    cookies.put("iPlanetDirectoryPro", successJson.get(TOKEN_ID).getAsString());
+
+    //  iPlanetDirectoryPro is OpenAM's login cookie. If it's lost it will log off.
+    GlobalSession.INSTANCE.addCookies(cookies);
+
+    String samlURL = queryParams.get("goto");
+
+    //  Send the SAMLResponse to the original location header
+    return RestAssured.given()
+        .cookies(cookies)
+        .queryParam("acsURL", queryParams.get("acsURL"))
+        .queryParam("binding", queryParams.get("binding"))
+        .queryParam("index", queryParams.get("index"))
+        .queryParam("ReqID", queryParams.get("ReqID"))
+        .queryParam("spEntityID", queryParams.get("spEntityID"))
```

Comment:
❓ Can you not just do `.queryParams(queryParams)`? Are there parameters we specifically need to leave out?

Reply:
Good question, I know that queryParams picks up more than it needs so I'll try it and see if it works. I was trying to mimic the request to the letter without having extra stuff in it.

Reply:
Tried it and it worked 🎉